### PR TITLE
Fix errors with JSON fields during resharding

### PR DESCRIPTION
**Pull request by deepthi**
reported by @acharis 
The testcase `TestPlayerCopyJsonData` reproduces an "Invalid value" error while trying to reshard tables that have json data in them.

```E0806 11:32:55.973881    2848 controller.go:134] stream 1: Invalid JSON text: "Invalid value." at position 0 in value for column 'src.val'. (errno 3140) (sqlstate 22032) during query: insert into src(id,val) select 1, 'JSON_OBJECT(\'aaa\',1)' from dual where (1) <= (2), retrying after 5s
--- FAIL: TestPlayerCopyJsonData (5.71s)
    vcopier_test.go:467: query:
        "insert into src(id,val) select 1, 'JSON_OBJECT(\\'aaa\\',1)' from dual where (1) <= (2)", does not match query 13:
        "insert into src(id,val) select 1, JSON_OBJECT('aaa',1) from dual where (1) <= (2)"
```
The problem is that the call to `JSON_OBJECT` is quoted in the generated query. This is because we are calling `ParsedQuery.Append` which eventually calls `sqltypes.Value.EncodeSQL`. This value is of type JSON (query.Type_JSON) which means `isQuoted()` is true, hence when it is encoded it is enclosed in single-quotes.

There may be a couple of ways of addressing this:
1. change printJSONObject and printJSONString to use `{...}` form for json objects. In the above example, the json processing will then produce `{"aaa": 1}` and EncodeSQL will quote it so that it becomes `'{"aaa": 1}'`. Similarly, instead of `JSON_ARRAY('a', 'b')` we would produce `["a", "b"]` and it will be quoted as `'["a", "b"]'`
2. change EncodeSQL to recognize when a "value" is actually a function call and not quote it, or change the calling code to handle this condition.

Not being super familiar with the json processing code, I have no idea what the implications of 1. might be. As for 2, it isn't clear from the code how one would accomplish that.

@sougou thoughts? 
EDIT (10/25/19): What we ended up implementing is basically option 1 above with some additional changes as needed.

### CODE REVIEW

File: go/mysql/binlog_event_json.go [hunk 1]
```diff
@@ -114,27 +118,32 @@ func printJSONObject(data []byte, large bool, result *bytes.Buffer) error {
...
 		if i > 0 {
 			result.WriteByte(',')
 		}
-		keys[i].EncodeSQL(result)
-		result.WriteByte(',')
+		keydata := keys[i]
+		//(data[keyOffset:keyOffset+keyLength])
+		result.WriteString("\"")
+		// FIXME(alainjobart): escape reserved characters
+		result.Write(data[keydata.offset : keydata.offset+keydata.length])
```
Comment (sougou) [inline go/mysql/binlog_event_json.go]: We have to test for `"` and other characters that json requires you to escape. I think they'll break. Same comments below for `printJSONString`.

File: go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go [hunk 2]
```diff
@@ -246,6 +246,267 @@ func TestPlayerCopyBigTable(t *testing.T) {
 	})
 }
 
+// TestPlayerCopyWildcardRule ensures the copy-catchup back-and-forth loop works correctly
+// when the filter uses a wildcard rule
+func TestPlayerCopyWildcardRule(t *testing.T) {
```
Comment (sougou) [inline go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go]: This doesn't seem to be part of the PR. Was this a test we discussed adding?

File: go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go [hunk 3]
```diff
@@ -246,6 +246,267 @@ func TestPlayerCopyBigTable(t *testing.T) {
...
+	expectData(t, "src", [][]string{
+		{"1", "aaa"},
+		{"2", "bbb"},
+		{"3", "ccc"},
+	})
+}
+
+// TestPlayerCopyJsonData ensures the copy-catchup back-and-forth loop works correctly
+// when the filter uses a wildcard rule & the tables have json data
+func TestPlayerCopyJsonData(t *testing.T) {
```
Comment (sougou) [inline go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go]: We should split this into two tests, one for copy and the other for vreplication. The back-and-forth part doesn't need to be combined with this.

File: go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go (hunk 2 above)
Comment (deepthi) [reply go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go]: I had the lastpk change on this branch because that PR had not been merged yet. Will rebase.

Comment (deepthi) [reply go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go]: done.

File: go/mysql/binlog_event_json.go (hunk 1 above)
Comment (deepthi) [reply go/mysql/binlog_event_json.go]: Since this is already a problem with the existing implementation, can we defer the handling of reserved characters? As it stands today, basic resharding with vreplication fails until this PR is merged.

File: go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go (hunk 3 above)
Comment (deepthi) [reply go/vt/vttablet/tabletmanager/vreplication/vcopier_test.go]: Done.

**Comment by morgo** [comment]
@deepthi ping

**Comment by deepthi** [comment]
I had discussed this with @sougou. We still have one unsolved problem which is escaping of special characters. I have not pursued that because it is not currently a high priority.
We can close this PR for now and re-open when we are ready to tackle the special character issue.
